### DISCUSSION TITLE
pythonic container interfaces

### DISCUSSION THREAD

**Discussion by mandersogit** [pyodide/pyodide#2903] [Ideas]
Has it been considered to present pythonic container interfaces for javascript container data structures where applicable?

In the 2022 Pycon Pyodide talk, there was an example of using `random.sample()` to get random elements from a javascript Array, and the need to first convert the Array into a python list to meet the expectations of `random.sample()`. I don't think this would be required if the JsProxy presented the `collections.abc.MutableSequence` (or `collections.abc.Sequence`) interface when proxying for an Array. 

One can construct a wrapper around the Array like this:

```python
from collections.abc import Sequence

class JsArraySequence(Sequence):

    def __init__(self, js_proxy):
        self.js_proxy = js_proxy

    def __getitem__(self, index):
        if isinstance(index, slice):
            indexes = range(*index.indices(len(self)))
            return tuple(self[idx] for idx in indexes)
        elif isinstance(index, int):
            if index < 0:
                index += len(self)
            if index < 0 or index >= len(self):
                raise IndexError(f"{index} out of bounds")
            value = self.js_proxy[index]
            return value

    def __len__(self):
        return self.js_proxy.length

    def __repr__(self):
        clsname = self.__class__.__name__
        return f"{clsname}({list(self)})"
```

Accessing the Array through this wrapper allows python code to work access the underlying Array almost as if it were a tuple. Implementing the abstract methods of MutableSequence (methods `__setitem__()`, `__delitem__()`, and `insert()`) allows python code to access and manipulate the Array almost as if it were a list.

The `random.sample()` function can work on the Array through this proxy, directly, without making a copy.

It seems to me like it would be very pythonic for the JsProxy to present a full MutableSequence interface for a JavaScript Array (probably in addition to the javascript Array API) and a full MutableMapping interface for a Javascript Object (and Map). Has this been considered?

If implemented in the JsProxy C code, I would think it would be very efficient.

**Comment by hoodmane** [comment]
Well we are already implementing `__getitem__` and `__setitem__` and `__len__`.

There are at least two different ways to check if something is a sequence: `PySequence_Check` is often used in C code and it just checks whether `tp_as_sequence->sq_item` is provided. This returns false for us because we have instead provided `tp_as_mapping->mp_subscript`. I think the reason we provide `mp_subscript` is that it handles slices efficiently, but `sq_item` doesn't. It seems that `listobject.c` defines both `mp_subscript` and `sq_item`. This means that `PySequence_Check` returns `True` but also the `mp_subscript` is what's used when you index a list. I guess we should follow their lead. 

When `random.sample` tests the `population` argument for being a list, it uses `isinstance(population, Sequence)` which is not the same as `PySequence_Check`. I think for this we may either have to set `Py_TPFLAGS_SEQUENCE` or call `MutableSequence.register(JsProxy_ArrayClass)` or something like that.

**Reply by hoodmane** [reply]
See #2906. It seems to work.

**Reply by mandersogit** [reply]
Awesome. That was fast.  Much appreciated.

It looks like a couple tests failed; does that stop the commit from being deployed into "dev"? When should I be able to see the change if I'm running from the CDN instead of building myself?

The API that I see on the current JsProxy array is missing:
```python
for attr in dir(MutableSequence):
  if not hasattr(js.arr, attr):
    print(attr)

#__abstractmethods__
#__class_getitem__
__iadd__
__reversed__
#__slots__
#_abc_impl
append
clear          # not on your list in #2906 
count
extend
index
insert         # not on your list in #2906
remove
```
You mention needing to add `pop()`, but a method named that is already there and at a glance does something reasonable.

I would advocate for implementing the entire interface, especially if there aren't any API name collisions between the javascript Array methods and the python list/Sequence methods. 

Would it be straightforward to support slice indexes in the `__getitem__()` method, or is that complicated for some reason? I imagine supporting slice indexes in `__setitem__()` might be quite complicated (I've created many MutableSequence subclasses over the years and never attempted it... `foo[0:10] = [1, 2]` is a little hairy to think about).

Were you also considering adding the python MutableMapping API to the proxy when the proxied object is a Javascript Map or Object? I'd very much like to see that as well.

**Reply by hoodmane** [reply]
> a couple tests failed

These don't look important. 

> When should I be able to see the change if I'm running from the CDN instead of building myself?

Only after it is merged into main. If you don't want to build it yourself you can download the build copy from the circleci artifacts. For the most recent commit that is here:
https://output.circle-artifacts.com/output/job/8d0a3849-bc9a-4c70-a109-3703880b133c/artifacts/0/root/repo/pyodide-build.tar.gz

> needing to add `pop()`

The Python `pop` takes an argument which specifies an index to remove. The JavaScript `pop` does not.

> Were you also considering adding the python MutableMapping API to the proxy when the proxied object is a Javascript Map or Object?

Yes to `Map`, no to `Object`. We could consider eventually adding an adaptor for `Object` but by default we do not define `__getitem__` on them unless they have a `get` method or they are an `Array`.

> The API that I see on the current JsProxy array is missing:

It's a fair amount of busy work to implement and test all of these. I probably won't get to it for a while. If you are willing to contribute some of these it would get done faster and we would appreciate it. I can provide advice.

> Would it be straightforward to support slice indexes in the __getitem__() method

Yeah if the slice is contiguous (`step` is `None` or `1`) then you can just use the JavaScript `.slice` method. If `step` is not 1 then you can use something like:

```js
if(start < 0){
	start = start + array.length;
}
if(stop === pyNone) {
	stop = array.length;
}
if(stop < 0) {
	stop = stop + array.length;
}
length = (stop - start + step - 1)/step;
result = Array.from({length}, (_,i)=> array[start + i*step]);
```

If the `step` is 1, a slice in `__setitem__` can be managed quite simply with [splice](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice). Otherwise I would probably figure out which elements need to be deleted, walk backwards across the list deleting them, then do replacements in a separate pass.

**Reply by mandersogit** [reply]
I would like to get involved and make contributions along these lines. I have lots of python experience, but little experience with a few of the other critical pieces of the puzzle (using or building emscripten, using node, pull requests on github, etc). I need to get to the point I can build and test pyodide before I can attempt any changes.

Is it possible to get there with only conda-installed packages, on linux or mac? The docs say to "Make sure the prerequisites for emsdk are installed", but it wasn't immediately obvious to me what those are. I only spent a few minutes working on it though.

I'll try to read up and figure it out how to get a build environment working somewhere, probably on my personal Mac.  If you have suggestions beyond what is in the pyodide docs, I'd love to hear them.

**Reply by hoodmane** [reply]
I usually build inside of the pyodide-env docker image since that is more convenient and reproducible. We usually use the `run_docker` script in the repo root directory for convenience. A Linux works best, people are building Pyodide on Mac but sometimes they run into weird issues and the docker image may be slow on the m1.

I think the docs on this are in decent shape though we are always trying to improve them so if you find anything unclear we are interested in pull requests or feedback.

**Reply by hoodmane** [reply]
One step that is needed currently (I think) that is missing from the docs is to use `pip install ./pyodide-test-runner` inside of the docker image before you can run tests.

**Reply by mandersogit** [reply]
We don't use docker in the datacenter at work (the tooling is not installed, and I'm not an admin), and I pre-date docker and haven't ever spent the time to get up to speed on it elsewhere. My understanding is that it kind of works on a Mac but not very well (I could be wrong). I don't currently have a linux host where I am admin where I could play with it.

I installed tooling on my Mac and tried to build pyodide core (what was on main when I cloned the repo this morning, no changes from me). It got pretty far I think, but it is failing with:

```
% env PATH=/usr/local/conda/envs/pyodide/bin:$PATH make
./tools/dependency-check.sh
npx rollup -c src/js/rollup.config.js
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating

./src/js/pyodide.ts → dist/pyodide.mjs...
[!] (plugin Typescript) TS1134: Variable declaration expected.
src/js/pyproxy.gen.ts (52:13)

52 declare var (1 << 8): number;
               ~

make: *** [src/js/_pyodide.out.js] Error 1
```

Telling `npx` to `--update-db` errors out without changing anything:
```
% env PATH=/usr/local/conda/envs/pyodide/bin:$PATH /usr/local/conda/envs/pyodide/bin/npx browserslist@latest --update-db

/Users/manderso/git/pyodide/src/js/node_modules/browserslist/update-db.js:15
    throw new BrowserslistError(
    ^
Error [BrowserslistError]: Cannot find package.json. Is this the right directory to run `npx browserslist --update-db` in?
    at detectLockfile (/Users/manderso/git/pyodide/src/js/node_modules/browserslist/update-db.js:15:11)
    ...
```

Is there a build environment elsewhere that I could leverage in theory? I'd like to finish getting it working on my local machine, but...

Is this a conversation that would be better taken off github discussions? Or is it good that there is an artifact with someone struggling to get up to speed for future users? ;-)

**Reply by hoodmane** [reply]
Yeah we might want a dedicated build issues for the Mac topic maybe in the docs. The error you are currently hitting is due to an incompatibility between the gnu sed and the Mac sed. I think it's possible to fix by installing gnu sed on the Mac.

**Reply by mandersogit** [reply]
I had already installed gnu-sed with homebrew.

I expected this to fix it:
```
% ln -s /usr/local/bin/gsed /usr/local/conda/envs/pyodide/bin/sed
% env PATH=/usr/local/conda/envs/pyodide/bin:$PATH make clean
rm -fr dist/*
% env PATH=/usr/local/conda/envs/pyodide/bin:$PATH make
./tools/dependency-check.sh
emcc -o src/core/docstring.o -c src/core/docstring.c -O2 -g0 -fPIC  -Wall -Wno-warn-absolute-paths -Werror=unused-variable -Werror=sometimes-uninitialized -Werror=int-conversion -Werror=incompatible-pointer-types -Werror=unused-result -I/Users/manderso/git/pyodide/cpython/installs/python-3.10.2/include/python3.10 -s EXCEPTION_CATCHING_ALLOWED=['we only want to allow exception handling in side modules'] -Isrc/core/
...
./src/js/pyodide.ts → dist/pyodide.mjs...
[!] (plugin Typescript) TS1134: Variable declaration expected.
src/js/pyproxy.gen.ts (52:13)

52 declare var (1 << 8): number;
               ~

make: *** [src/js/_pyodide.out.js] Error 1
```
 but it didn't.

Is there an intermediate file that needs to get deleted that `make clean` does not delete?

**Reply by hoodmane** [reply]
Probably `src/js/pyproxy.gen.ts`?

**Reply by mandersogit** [reply]
Deleted `src/js/pyproxy.gen.ts` (which _I_ think should have been deleted by `make clean`) and got quite a bit further again.  Now we are failing with

```
Error building fpcast-test. Printing build logs.py, micropip
Traceback (most recent call last):
  File "/usr/local/conda/envs/pyodide/lib/python3.10/site-packages/pep517/in_process/_in_process.py", line 363, in <module>
    main()
...
  File "/private/var/folders/29/vh0lhg2n2bvbnycdh00bk0r80000gn/T/build-env-qdq_wq91/lib/python3.10/site-packages/setuptools/build_meta.py", line 174, in run_setup
    exec(code, locals())
  File "<string>", line 9, in <module>
Exception:
This should appear in the log exactly one time. If it appears more than once,
the Pyodide build system has misconfigured sysconfigdata (and also the build
will fail).
...
ERROR Backend subproccess exited when trying to invoke build_wheel
[2022-07-28 16:25:11] Succeeded building package fpcast-test in 19.6 seconds.
ERROR: cancelling buildall
Fatal Python error: _enter_buffered_busy: could not acquire lock for <_io.BufferedWriter name='<stdout>'> at interpreter shutdown, possibly due to daemon threads
Python runtime state: finalizing (tstate=0x00007ff098704080)

Current thread 0x000000010d969600 (most recent call first):
  <no Python frame>

Extension modules: _ruamel_yaml, yaml._yaml (total: 2)
make[1]: *** [all] Abort trap: 6
make: *** [dist/repodata.json] Error 2
```

I didn't really think it would be relevant, but this doesn't seem to work:

```
% /usr/local/conda/envs/pyodide/bin/pip install ./pyodide-test-runner/
ERROR: Directory './pyodide-test-runner/' is not installable. Neither 'setup.py' nor 'pyproject.toml' found.
% /usr/local/conda/envs/pyodide/bin/pip install ./pyodide-test-runner/pyodide_test_runner
ERROR: Directory './pyodide-test-runner/pyodide_test_runner' is not installable. Neither 'setup.py' nor 'pyproject.toml' found.
```
And it is correct that neither `setup.py` nor `pyproject.toml` are in that folder.

**Reply by hoodmane** [reply]
> I didn't really think it would be relevant

Yeah that isn't relevant to the current failure. But there should be a `pyproject.toml` file in `pyodide-test-runner`:
https://github.com/pyodide/pyodide/blob/main/pyodide-test-runner/pyproject.toml

**Reply by hoodmane** [reply]
> Deleted `src/js/pyproxy.gen.ts` (which I think should have been deleted by make clean)

Yes it should have. We should add `rm -f src/js/*.gen.ts` and `src/js/_pyodide.out.js` to the `make clean` recipe.

**Reply by hoodmane** [reply]
Could you include more of the log from the build failure? I think the `...`s contain something important because the error you did print is expected (hence `This should appear in the log exactly one time.`) and `Fatal Python error:` is a common side effect of a build error caused by something else but isn't diagnostic of a problem by itself.

**Reply by mandersogit** [reply]
```
make -C packages
/usr/local/conda/envs/pyodide/bin/python3.10 -m pip install -e ../pyodide-build
Obtaining file:///Users/manderso/git/pyodide/pyodide-build
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: pyyaml in /usr/local/conda/envs/pyodide/lib/python3.10/site-packages (from pyodide-build==0.21.0.dev0) (6.0)
Requirement already satisfied: wheel in /usr/local/conda/envs/pyodide/lib/python3.10/site-packages (from pyodide-build==0.21.0.dev0) (0.37.1)
Requirement already satisfied: build==0.7.0 in /usr/local/conda/envs/pyodide/lib/python3.10/site-packages (from pyodide-build==0.21.0.dev0) (0.7.0)
Requirement already satisfied: packaging in /usr/local/conda/envs/pyodide/lib/python3.10/site-packages (from pyodide-build==0.21.0.dev0) (21.3)
Requirement already satisfied: ruamel.yaml in /usr/local/conda/envs/pyodide/lib/python3.10/site-packages (from pyodide-build==0.21.0.dev0) (0.17.21)
Requirement already satisfied: cython<3.0 in /usr/local/conda/envs/pyodide/lib/python3.10/site-packages (from pyodide-build==0.21.0.dev0) (0.29.31)
Requirement already satisfied: tomli in /usr/local/conda/envs/pyodide/lib/python3.10/site-packages (from pyodide-build==0.21.0.dev0) (2.0.1)
Requirement already satisfied: pep517>=0.9.1 in /usr/local/conda/envs/pyodide/lib/python3.10/site-packages (from build==0.7.0->pyodide-build==0.21.0.dev0) (0.12.0)
Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /usr/local/conda/envs/pyodide/lib/python3.10/site-packages (from packaging->pyodide-build==0.21.0.dev0) (3.0.9)
Requirement already satisfied: ruamel.yaml.clib>=0.2.6 in /usr/local/conda/envs/pyodide/lib/python3.10/site-packages (from ruamel.yaml->pyodide-build==0.21.0.dev0) (0.2.6)
Installing collected packages: pyodide-build
  Attempting uninstall: pyodide-build
    Found existing installation: pyodide-build 0.21.0.dev0
    Uninstalling pyodide-build-0.21.0.dev0:
      Successfully uninstalled pyodide-build-0.21.0.dev0
  Running setup.py develop for pyodide-build
Successfully installed pyodide-build-0.21.0.dev0
mkdir -p /Users/manderso/git/pyodide/packages/.artifacts /Users/manderso/git/pyodide/packages/.libs
PYODIDE_ROOT=/Users/manderso/git/pyodide python -m pyodide_build buildall . /Users/manderso/git/pyodide/dist \
	 --n-jobs ${PYODIDE_JOBS:-4} \
	--log-dir=./build-logs
The following packages are already built: MarkupSafe, atomicwrites, attrs, cpp-exceptions-test, distutils, iniconfig, more-itertools, openssl, packaging, pluggy, py, pyparsing, setuptools, sharedlib-test, six, and test

Building the following packages: Jinja2, fpcast-test, micropip, pytest, pytz, regex, sharedlib-test-py, ssl, and tblib
[1/9] (thread 1) building pytest
[2/9] (thread 2) building fpcast-test
[3/9] (thread 3) building pytzst
[4/9] (thread 4) building tblib
[4/9] (thread 4) built tblib in 2.84 s
[5/9] (thread 4) building regex
[1/9] (thread 1) built pytest in 5.45 s
[6/9] (thread 1) building ssl
[3/9] (thread 3) built pytz in 7.15 s
[7/9] (thread 3) building sharedlib-test-py
[6/9] (thread 1) built ssl in 5.52 s
[8/9] (thread 1) building Jinja2
[8/9] (thread 1) built Jinja2 in 3.06 s
[9/9] (thread 1) building micropip
Error building fpcast-test. Printing build logs.py, micropip
Traceback (most recent call last):
  File "/usr/local/conda/envs/pyodide/lib/python3.10/site-packages/pep517/in_process/_in_process.py", line 363, in <module>
    main()
  File "/usr/local/conda/envs/pyodide/lib/python3.10/site-packages/pep517/in_process/_in_process.py", line 345, in main
    json_out['return_val'] = hook(**hook_input['kwargs'])
  File "/usr/local/conda/envs/pyodide/lib/python3.10/site-packages/pep517/in_process/_in_process.py", line 130, in get_requires_for_build_wheel
    return hook(config_settings)
  File "/private/var/folders/29/vh0lhg2n2bvbnycdh00bk0r80000gn/T/build-env-qdq_wq91/lib/python3.10/site-packages/setuptools/build_meta.py", line 177, in get_requires_for_build_wheel
    return self._get_build_requires(
  File "/private/var/folders/29/vh0lhg2n2bvbnycdh00bk0r80000gn/T/build-env-qdq_wq91/lib/python3.10/site-packages/setuptools/build_meta.py", line 159, in _get_build_requires
    self.run_setup()
  File "/private/var/folders/29/vh0lhg2n2bvbnycdh00bk0r80000gn/T/build-env-qdq_wq91/lib/python3.10/site-packages/setuptools/build_meta.py", line 281, in run_setup
    super(_BuildMetaLegacyBackend,
  File "/private/var/folders/29/vh0lhg2n2bvbnycdh00bk0r80000gn/T/build-env-qdq_wq91/lib/python3.10/site-packages/setuptools/build_meta.py", line 174, in run_setup
    exec(code, locals())
  File "<string>", line 9, in <module>
Exception:
This should appear in the log exactly one time. If it appears more than once,
the Pyodide build system has misconfigured sysconfigdata (and also the build
will fail).

running egg_info
creating fpcast_test.egg-info
writing fpcast_test.egg-info/PKG-INFO
writing dependency_links to fpcast_test.egg-info/dependency_links.txt
writing top-level names to fpcast_test.egg-info/top_level.txt
writing manifest file 'fpcast_test.egg-info/SOURCES.txt'
reading manifest file 'fpcast_test.egg-info/SOURCES.txt'
writing manifest file 'fpcast_test.egg-info/SOURCES.txt'
running bdist_wheel
running build
running build_ext
building 'fpcast_test' extension
creating build
creating build/temp.emscripten_3_1_14_wasm32-cpython-310
cc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -O2 -g0 -fPIC -I/private/var/folders/29/vh0lhg2n2bvbnycdh00bk0r80000gn/T/build-env-qdq_wq91/include -I/usr/local/conda/envs/pyodide/include/python3.10 -c fpcast-test.c -o build/temp.emscripten_3_1_14_wasm32-cpython-310/fpcast-test.o
/Users/manderso/git/pyodide/emsdk/emsdk/upstream/emscripten/emcc: line 39: exec: ccache: not found
error: command '/Users/manderso/git/pyodide/tools/symlinks/cc' failed with exit code 127
[2022-07-28 16:24:51] Building package fpcast-test...
* Creating virtualenv isolated environment...
* Installing packages in isolated environment... (setuptools >= 40.8.0, wheel)
* Installing packages in isolated environment... (cython, pythran)
* Getting dependencies for wheel...
* Getting dependencies for wheel...
* Installing packages in isolated environment... (wheel)
* Installing packages in isolated environment... (cython, pythran)
* Building wheel...

ERROR Backend subproccess exited when trying to invoke build_wheel
[2022-07-28 16:25:11] Succeeded building package fpcast-test in 19.6 seconds.
ERROR: cancelling buildall
Fatal Python error: _enter_buffered_busy: could not acquire lock for <_io.BufferedWriter name='<stdout>'> at interpreter shutdown, possibly due to daemon threads
Python runtime state: finalizing (tstate=0x00007ff098704080)

Current thread 0x000000010d969600 (most recent call first):
  <no Python frame>

Extension modules: _ruamel_yaml, yaml._yaml (total: 2)
make[1]: *** [all] Abort trap: 6
make: *** [dist/repodata.json] Error 2
```

**Reply by hoodmane** [reply]
I think the actual failure there is most likely to be this line:
```
/Users/manderso/git/pyodide/emsdk/emsdk/upstream/emscripten/emcc: line 39: exec: ccache: not found
```
though I am not sure what to make of it.

**Reply by hoodmane** [reply]
Building `fpcast-test` isn't very necessary though. You could try adding `_disabled: true` to line 4 of `packages/fpcast-test/meta.yaml` and see if the other packages build successfully.

**Reply by hoodmane** [reply]
You might just want to say `PYODIDE_PACKAGES='micropip,tblib,pytest'` because I think those are the only packages that are relevant for testing changes to `src/core/jsproxy.c` and friends.

**Reply by mandersogit** [reply]
I didn't have a chance to come back to this for a few days.

Disabling `fpcast-test` didn't help. But doing a `brew install ccache` did install a package `ccache` and that allowed the make process to complete to a point where it said "SUCCESS!" (haven't otherwise tested it).

I built from a fresh clone of the repo and it built to the same "SUCCESS!" point, with no interventions. It appears that `ccache` is currently a required package on the Mac, or something is configured wrong if this shouldn't be the case.

I'm not going to be able to circle back to this for at least a couple of weeks at this point (traveling). But I look forward to trying to dabble with a feature enhancement. Doesn't need to be this specifically -- if you have time to work on this without my involvement, please go ahead.

**Comment by rth** [comment]
The other relevant discussion is in [NEP 47 — Adopting the array API standard](https://numpy.org/neps/nep-0047-array-api-standard.html), 

>  Implementing the abstract methods of MutableSequence (methods __setitem__(), __delitem__(), and insert()) allows python code to access and manipulate the Array almost as if it were a list.

An array is rather different from a list though, in the sense that it's a contiguous block of memory and deleting an element or inserting a new one needs copying/re-allocating everything which is really inefficient (and users should be very aware of that). [array.array](https://docs.python.org/3/library/array.html) does indeed implement delitem etc, but I feel we should rather go for compatibility with numpy. And numpy does not implement MutableSequence methods cf https://github.com/pyodide/pyodide/pull/2906#issuecomment-1203842014 so I'm not sure that is the right direction.

Maybe we should talk to numpy people to see what's the easiest / best way to achieve API compatibility.

**Reply by hoodmane** [reply]
Well the JavaScript `Array` is similar to a Python `list` in functionality and in efficiency whereas a `TypedArray` is much more similar to a Python `array`. So I think going ahead and implementing the `list` methods on JavaScript Arrays makes a lot of sense.

**Reply by rth** [reply]
Yeah, my comment was more about TypedArray, for Array you are right, it makes sense to do it.

**Reply by hoodmane** [reply]
Right I haven't checked recently what we are doing with typed arrays. But I guess they should also be `MutableSequence`s then?

**Reply by rth** [reply]
I'm not sure about TypedArray; we should ask the numpy community.

**Reply by hoodmane** [reply]
I guess currently we treat `Array` and `TypedArray` identically: `JsArray_Check` returns true for both. Probably we should not do that. For instance, when slicing a `TypedArray` we probably want to use `subarray` which produces a view as opposed to `slice` which is a copy. This is made slightly more complicated though by the possibility of slicing with a step not equal to 1...
